File: leet-code/wordBreak.js
/**
 * @param {string} s
 * @param {string[]} wordDict
 * @return {string[]}
 */
var wordBreak = function (s, wordDict) {
  const map = new Map()

  let dfs = function (s, len, wordSet, index, map) {
    console.log(index, map)
    if (map.has(index)) {
      return map.get(index)
    }
    const wordBreaks = []
    if (index === len) {
      wordBreaks.push([])
    }

    for (let i = index + 1; i <= len; i++) {
      const word = s.substring(index, i)
      if (wordSet.has(word)) {
        const nextWordBreaks = dfs(s, len, wordSet, i, map)
        for (const item of nextWordBreaks) {
          const wordBreak = [word, ...item]
          wordBreaks.push(wordBreak)
        }
      }
    }
    map.set(index, wordBreaks)
    return wordBreaks
  }

  const wordBreaks = dfs(s, s.length, new Set(wordDict), 0, map)
  const res = []
  for (const item of wordBreaks) {
    res.push(item.join(' '))
  }
  return res
}

console.log(wordBreak('catsanddog', ['cat', 'cats', 'and', 'sand', 'dog']))

File: leet-code/wordBreak.go
package main

import (
	"fmt"
	"strings"
)

func wordBreak(s string, wordDict []string) []string {
	wordSet := make(map[string]struct{})
	wordMap := make(map[int][][]string)
	res := make([]string, 0)
	for _, c := range wordDict {
		wordSet[c] = struct{}{}
	}

	var dfs func(s string, wordSet map[string]struct{}, index int, wordMap map[int][][]string) [][]string
	dfs = func(s string, wordSet map[string]struct{}, index int, wordMap map[int][][]string) [][]string {
		fmt.Println(111, wordSet)
		if _, has := wordMap[index]; has {
			return wordMap[index]
		}
		fmt.Println(222)

		wordBreaks := make([][]string, 0)
		if index == len(s) {
			var arr []string
			wordBreaks = append(wordBreaks, arr)
		}
		fmt.Println(333, len(s))

		for i := index + 1; i <= len(s); i++ {
			word := s[index:i]
			fmt.Println(444, s, i, word)
			if _, has := wordSet[word]; has {
				nextWordBreaks := dfs(s, wordSet, i, wordMap)
				for i := range nextWordBreaks {
					nextWordBreak := make([]string, 0)
					nextWordBreak = append(nextWordBreak, word)
					nextWordBreak = append(nextWordBreak, nextWordBreaks[i]...)
					wordBreaks = append(wordBreaks, nextWordBreak)
				}
			}
		}
		fmt.Println(555)

		wordMap[index] = wordBreaks
		return wordBreaks
	}
	wordBreaks := dfs(s, wordSet, 0, wordMap)
	for _, wordBreak := range wordBreaks {
		res = append(res, strings.Join(wordBreak, " "))
	}
	return res
}
